File: lua/tests/test_all.lua
-- Comprehensive tests for DSA library

package.path = package.path .. ";../?.lua;../data_structures/?.lua;../algorithms/?.lua"

local Stack = require("data_structures.stack")
local Queue = require("data_structures.queue")
local DynamicArray = require("data_structures.dynamic_array")
local SinglyLinkedList = require("data_structures.singly_linked_list")
local DoublyLinkedList = require("data_structures.doubly_linked_list")
local Deque = require("data_structures.deque")
local HashTable = require("data_structures.hash_table")
local BinarySearchTree = require("data_structures.binary_search_tree")
local MinHeap = require("data_structures.min_heap")
local DisjointSet = require("data_structures.disjoint_set")
local binary_search = require("algorithms.binary_search")
local insertion_sort = require("algorithms.insertion_sort")
local merge_sort = require("algorithms.merge_sort")
local quick_sort = require("algorithms.quick_sort")
local Graph = require("algorithms.graph")
local bfs = require("algorithms.bfs")
local dfs = require("algorithms.dfs")

local function assertEquals(expected, actual, msg)
    if expected ~= actual then
        error(string.format("%s: expected %s, got %s", msg or "Assertion failed", tostring(expected), tostring(actual)))
    end
end

local function assertArrayEquals(expected, actual, msg)
    if #expected ~= #actual then
        error(string.format("%s: arrays have different lengths", msg or "Assertion failed"))
    end
    for i = 1, #expected do
        if expected[i] ~= actual[i] then
            error(string.format("%s: arrays differ at index %d", msg or "Assertion failed", i))
        end
    end
end

local function assertTrue(value, msg)
    if not value then
        error(msg or "Expected true")
    end
end

local function assertFalse(value, msg)
    if value then
        error(msg or "Expected false")
    end
end

-- Stack Tests
print("Testing Stack...")
local s = Stack.new()
assertTrue(s:isEmpty())
s:push(1)
s:push(2)
s:push(3)
assertFalse(s:isEmpty())
assertEquals(3, s:size())
assertEquals(3, s:pop())
assertEquals(2, s:peek())
assertEquals(2, s:size())
print("Stack tests passed!")

-- Queue Tests
print("Testing Queue...")
local q = Queue.new()
assertTrue(q:isEmpty())
q:enqueue(1)
q:enqueue(2)
q:enqueue(3)
assertFalse(q:isEmpty())
assertEquals(3, q:size())
assertEquals(1, q:dequeue())
assertEquals(2, q:peek())
assertEquals(2, q:size())
print("Queue tests passed!")

-- DynamicArray Tests
print("Testing DynamicArray...")
local arr = DynamicArray.new()
arr:push(1)
arr:push(2)
arr:push(3)
assertEquals(1, arr:get(1))
assertEquals(2, arr:get(2))
arr:set(2, 5)
assertEquals(5, arr:get(2))
assertEquals(3, arr:pop())
assertEquals(2, arr:size())
arr:insert(2, 10)
assertEquals(10, arr:get(2))
assertEquals(10, arr:removeAt(2))
print("DynamicArray tests passed!")

-- SinglyLinkedList Tests
print("Testing SinglyLinkedList...")
local sll = SinglyLinkedList.new()
sll:append(2)
sll:prepend(1)
sll:append(3)
assertEquals(1, sll:get(1))
assertEquals(2, sll:get(2))
assertEquals(3, sll:get(3))
assertEquals(3, sll:size())
assertEquals(1, sll:removeFirst())
assertEquals(2, sll:size())
print("SinglyLinkedList tests passed!")

-- DoublyLinkedList Tests
print("Testing DoublyLinkedList...")
local dll = DoublyLinkedList.new()
dll:append(2)
dll:prepend(1)
dll:append(3)
assertEquals(1, dll:get(1))
assertEquals(2, dll:get(2))
assertEquals(3, dll:get(3))
assertEquals(1, dll:removeFirst())
assertEquals(3, dll:removeLast())
assertEquals(1, dll:size())
print("DoublyLinkedList tests passed!")

-- Deque Tests
print("Testing Deque...")
local d = Deque.new()
d:pushBack(2)
d:pushFront(1)
d:pushBack(3)
assertEquals(1, d:peekFront())
assertEquals(3, d:peekBack())
assertEquals(1, d:popFront())
assertEquals(3, d:popBack())
assertEquals(1, d:size())
print("Deque tests passed!")

-- HashTable Tests
print("Testing HashTable...")
local ht = HashTable.new()
ht:put("one", 1)
ht:put("two", 2)
assertEquals(1, ht:get("one"))
assertEquals(2, ht:get("two"))
assertTrue(ht:contains("one"))
assertFalse(ht:contains("three"))
assertTrue(ht:remove("one"))
assertFalse(ht:contains("one"))
print("HashTable tests passed!")

-- BinarySearchTree Tests
print("Testing BinarySearchTree...")
local bst = BinarySearchTree.new()
bst:insert(5)
bst:insert(3)
bst:insert(7)
bst:insert(1)
bst:insert(9)
assertTrue(bst:contains(5))
assertTrue(bst:contains(3))
assertFalse(bst:contains(6))
assertEquals(1, bst:findMin())
assertEquals(9, bst:findMax())
assertArrayEquals({1, 3, 5, 7, 9}, bst:inorder())
assertTrue(bst:remove(3))
assertFalse(bst:contains(3))
print("BinarySearchTree tests passed!")

-- MinHeap Tests
print("Testing MinHeap...")
local heap = MinHeap.new()
heap:insert(5)
heap:insert(3)
heap:insert(7)
heap:insert(1)
assertEquals(1, heap:peek())
assertEquals(1, heap:extractMin())
assertEquals(3, heap:extractMin())
assertEquals(5, heap:extractMin())
assertEquals(7, heap:extractMin())

local heap2 = MinHeap.heapify({5, 3, 7, 1, 9})
assertEquals(1, heap2:peek())
print("MinHeap tests passed!")

-- DisjointSet Tests
print("Testing DisjointSet...")
local ds = DisjointSet.new()
ds:makeSet(1)
ds:makeSet(2)
ds:makeSet(3)
ds:makeSet(4)
assertFalse(ds:connected(1, 2))
ds:union(1, 2)
assertTrue(ds:connected(1, 2))
ds:union(3, 4)
assertTrue(ds:connected(3, 4))
assertFalse(ds:connected(1, 3))
ds:union(2, 3)
assertTrue(ds:connected(1, 4))
print("DisjointSet tests passed!")

-- BinarySearch Tests
print("Testing BinarySearch...")
local sortedArr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
assertEquals(5, binary_search.search(sortedArr, 5))
assertEquals(1, binary_search.search(sortedArr, 1))
assertEquals(10, binary_search.search(sortedArr, 10))
assertEquals(nil, binary_search.search(sortedArr, 11))

local arr2 = {1, 2, 2, 2, 3, 4, 5}
assertEquals(2, binary_search.lowerBound(arr2, 2))
assertEquals(5, binary_search.upperBound(arr2, 2))
print("BinarySearch tests passed!")

-- InsertionSort Tests
print("Testing InsertionSort...")
local toSort = {5, 2, 8, 1, 9, 3}
insertion_sort.sort(toSort)
assertArrayEquals({1, 2, 3, 5, 8, 9}, toSort)

local toSort2 = {5, 2, 8, 1, 9, 3}
insertion_sort.sortDesc(toSort2)
assertArrayEquals({9, 8, 5, 3, 2, 1}, toSort2)
print("InsertionSort tests passed!")

-- MergeSort Tests
print("Testing MergeSort...")
local toSort3 = {5, 2, 8, 1, 9, 3, 7, 4, 6}
merge_sort.sort(toSort3)
assertArrayEquals({1, 2, 3, 4, 5, 6, 7, 8, 9}, toSort3)
print("MergeSort tests passed!")

-- QuickSort Tests
print("Testing QuickSort...")
local toSort4 = {5, 2, 8, 1, 9, 3, 7, 4, 6}
quick_sort.sort(toSort4)
assertArrayEquals({1, 2, 3, 4, 5, 6, 7, 8, 9}, toSort4)
print("QuickSort tests passed!")

-- Graph Tests
print("Testing Graph...")
local g = Graph.new(false)
g:addEdge(1, 2)
g:addEdge(1, 3)
assertTrue(g:hasVertex(1))
assertTrue(g:hasVertex(2))
assertTrue(g:hasEdge(1, 2))
assertTrue(g:hasEdge(2, 1))

local dg = Graph.new(true)
dg:addEdge(1, 2)
assertTrue(dg:hasEdge(1, 2))
assertFalse(dg:hasEdge(2, 1))
print("Graph tests passed!")

-- BFS Tests
print("Testing BFS...")
local g2 = Graph.new(false)
g2:addEdge(1, 2)
g2:addEdge(1, 3)
g2:addEdge(2, 4)

local order = bfs.traverse(g2, 1)
assertEquals(1, order[1])
assertEquals(4, #order)

local path = bfs.shortestPath(g2, 1, 4)
assertEquals(1, path[1])
assertEquals(4, path[#path])
assertEquals(3, #path)
print("BFS tests passed!")

-- DFS Tests
print("Testing DFS...")
local g3 = Graph.new(false)
g3:addEdge(1, 2)
g3:addEdge(1, 3)
g3:addEdge(2, 4)

local order2 = dfs.traverse(g3, 1)
assertEquals(1, order2[1])
assertEquals(4, #order2)

local cyclic = Graph.new(true)
cyclic:addEdge(1, 2)
cyclic:addEdge(2, 3)
cyclic:addEdge(3, 1)
assertTrue(dfs.hasCycle(cyclic))

local acyclic = Graph.new(true)
acyclic:addEdge(1, 2)
acyclic:addEdge(2, 3)
assertFalse(dfs.hasCycle(acyclic))

local sorted = dfs.topologicalSort(acyclic)
assertEquals(3, #sorted)
print("DFS tests passed!")

print("")
print("=================================")
print("All tests passed!")
print("=================================")

File: nim/tests/test_all.nim
## Comprehensive tests for DSA library

import std/unittest
import std/sequtils
import ../src/dsa

suite "Stack Tests":
  test "push and pop":
    var s = newStack[int]()
    s.push(1)
    s.push(2)
    s.push(3)
    check s.pop() == 3
    check s.pop() == 2
    check s.pop() == 1

  test "peek":
    var s = newStack[int]()
    s.push(42)
    check s.peek() == 42
    check s.size() == 1

  test "isEmpty":
    var s = newStack[int]()
    check s.isEmpty()
    s.push(1)
    check not s.isEmpty()

suite "Queue Tests":
  test "enqueue and dequeue":
    var q = newQueue[int]()
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    check q.dequeue() == 1
    check q.dequeue() == 2
    check q.dequeue() == 3

  test "peek":
    var q = newQueue[int]()
    q.enqueue(42)
    check q.peek() == 42
    check q.size() == 1

suite "DynamicArray Tests":
  test "push and get":
    var arr = newDynamicArray[int]()
    arr.push(1)
    arr.push(2)
    arr.push(3)
    check arr[0] == 1
    check arr[1] == 2
    check arr[2] == 3

  test "set":
    var arr = newDynamicArray[int]()
    arr.push(1)
    arr[0] = 42
    check arr[0] == 42

  test "insert and removeAt":
    var arr = newDynamicArray[int]()
    arr.push(1)
    arr.push(3)
    arr.insert(1, 2)
    check arr[1] == 2
    check arr.removeAt(1) == 2
    check arr.size() == 2

suite "SinglyLinkedList Tests":
  test "prepend and append":
    var list = newSinglyLinkedList[int]()
    list.append(2)
    list.prepend(1)
    list.append(3)
    check list[0] == 1
    check list[1] == 2
    check list[2] == 3

  test "removeFirst":
    var list = newSinglyLinkedList[int]()
    list.append(1)
    list.append(2)
    check list.removeFirst() == 1
    check list.size() == 1

suite "DoublyLinkedList Tests":
  test "prepend and append":
    var list = newDoublyLinkedList[int]()
    list.append(2)
    list.prepend(1)
    list.append(3)
    check list[0] == 1
    check list[1] == 2
    check list[2] == 3

  test "removeFirst and removeLast":
    var list = newDoublyLinkedList[int]()
    list.append(1)
    list.append(2)
    list.append(3)
    check list.removeFirst() == 1
    check list.removeLast() == 3
    check list.size() == 1

suite "Deque Tests":
  test "pushFront and pushBack":
    var d = newDeque[int]()
    d.pushBack(2)
    d.pushFront(1)
    d.pushBack(3)
    check d.peekFront() == 1
    check d.peekBack() == 3

  test "popFront and popBack":
    var d = newDeque[int]()
    d.pushBack(1)
    d.pushBack(2)
    d.pushBack(3)
    check d.popFront() == 1
    check d.popBack() == 3

suite "HashTable Tests":
  test "put and get":
    var ht = newHashTable[string, int]()
    ht["one"] = 1
    ht["two"] = 2
    check ht["one"] == 1
    check ht["two"] == 2

  test "contains and remove":
    var ht = newHashTable[string, int]()
    ht["key"] = 42
    check ht.contains("key")
    check ht.remove("key")
    check not ht.contains("key")

suite "BinarySearchTree Tests":
  test "insert and contains":
    var tree = newBinarySearchTree[int]()
    tree.insert(5)
    tree.insert(3)
    tree.insert(7)
    check tree.contains(5)
    check tree.contains(3)
    check tree.contains(7)
    check not tree.contains(4)

  test "findMin and findMax":
    var tree = newBinarySearchTree[int]()
    tree.insert(5)
    tree.insert(3)
    tree.insert(7)
    tree.insert(1)
    tree.insert(9)
    check tree.findMin() == 1
    check tree.findMax() == 9

  test "remove":
    var tree = newBinarySearchTree[int]()
    tree.insert(5)
    tree.insert(3)
    tree.insert(7)
    check tree.remove(3)
    check not tree.contains(3)

  test "inorder":
    var tree = newBinarySearchTree[int]()
    tree.insert(5)
    tree.insert(3)
    tree.insert(7)
    tree.insert(1)
    tree.insert(9)
    check tree.inorder() == @[1, 3, 5, 7, 9]

suite "MinHeap Tests":
  test "insert and extractMin":
    var heap = newMinHeap[int]()
    heap.insert(5)
    heap.insert(3)
    heap.insert(7)
    heap.insert(1)
    check heap.extractMin() == 1
    check heap.extractMin() == 3
    check heap.extractMin() == 5
    check heap.extractMin() == 7

  test "peek":
    var heap = newMinHeap[int]()
    heap.insert(5)
    heap.insert(3)
    check heap.peek() == 3
    check heap.size() == 2

  test "heapify":
    var heap = heapify(@[5, 3, 7, 1, 9])
    check heap.extractMin() == 1

suite "DisjointSet Tests":
  test "makeSet and find":
    var ds = newDisjointSet[int]()
    ds.makeSet(1)
    ds.makeSet(2)
    check ds.find(1) == 1
    check ds.find(2) == 2

  test "union and connected":
    var ds = newDisjointSet[int]()
    ds.makeSet(1)
    ds.makeSet(2)
    ds.makeSet(3)
    check not ds.connected(1, 2)
    discard ds.union(1, 2)
    check ds.connected(1, 2)
    check not ds.connected(1, 3)

suite "BinarySearch Tests":
  test "binarySearch":
    let arr = @[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    check binarySearch(arr, 5) == 4
    check binarySearch(arr, 1) == 0
    check binarySearch(arr, 10) == 9
    check binarySearch(arr, 11) == -1

  test "lowerBound and upperBound":
    let arr = @[1, 2, 2, 2, 3, 4, 5]
    check lowerBound(arr, 2) == 1
    check upperBound(arr, 2) == 4

suite "InsertionSort Tests":
  test "sort":
    var arr = @[5, 2, 8, 1, 9, 3]
    insertionSort(arr)
    check arr == @[1, 2, 3, 5, 8, 9]

  test "sortDesc":
    var arr = @[5, 2, 8, 1, 9, 3]
    insertionSortDesc(arr)
    check arr == @[9, 8, 5, 3, 2, 1]

suite "MergeSort Tests":
  test "sort":
    var arr = @[5, 2, 8, 1, 9, 3, 7, 4, 6]
    mergeSort(arr)
    check arr == @[1, 2, 3, 4, 5, 6, 7, 8, 9]

  test "sortIterative":
    var arr = @[5, 2, 8, 1, 9, 3, 7, 4, 6]
    mergeSortIterative(arr)
    check arr == @[1, 2, 3, 4, 5, 6, 7, 8, 9]

suite "QuickSort Tests":
  test "sort":
    var arr = @[5, 2, 8, 1, 9, 3, 7, 4, 6]
    quickSort(arr)
    check arr == @[1, 2, 3, 4, 5, 6, 7, 8, 9]

  test "quickSelect":
    var arr = @[5, 2, 8, 1, 9]
    check quickSelect(arr, 0) == 1
    arr = @[5, 2, 8, 1, 9]
    check quickSelect(arr, 4) == 9

suite "Graph Tests":
  test "addEdge and hasEdge":
    var g = newGraph[int](directed = false)
    g.addEdge(1, 2)
    g.addEdge(1, 3)
    check g.hasEdge(1, 2)
    check g.hasEdge(2, 1)
    check g.hasEdge(1, 3)

  test "directed graph":
    var g = newGraph[int](directed = true)
    g.addEdge(1, 2)
    check g.hasEdge(1, 2)
    check not g.hasEdge(2, 1)

suite "BFS Tests":
  test "traverse":
    var g = newGraph[int](directed = false)
    g.addEdge(1, 2)
    g.addEdge(1, 3)
    g.addEdge(2, 4)
    let order = bfs.traverse(g, 1)
    check order[0] == 1
    check order.len == 4

  test "shortestPath":
    var g = newGraph[int](directed = false)
    g.addEdge(1, 2)
    g.addEdge(2, 3)
    g.addEdge(1, 3)
    let path = bfs.shortestPath(g, 1, 3)
    check path == @[1, 3]

suite "DFS Tests":
  test "traverse":
    var g = newGraph[int](directed = false)
    g.addEdge(1, 2)
    g.addEdge(1, 3)
    g.addEdge(2, 4)
    let order = dfs.traverse(g, 1)
    check order[0] == 1
    check order.len == 4

  test "hasCycle":
    var cyclic = newGraph[int](directed = true)
    cyclic.addEdge(1, 2)
    cyclic.addEdge(2, 3)
    cyclic.addEdge(3, 1)
    check dfs.hasCycle(cyclic)

    var acyclic = newGraph[int](directed = true)
    acyclic.addEdge(1, 2)
    acyclic.addEdge(2, 3)
    check not dfs.hasCycle(acyclic)

  test "topologicalSort":
    var g = newGraph[int](directed = true)
    g.addEdge(1, 2)
    g.addEdge(1, 3)
    g.addEdge(2, 3)
    let sorted = dfs.topologicalSort(g)
    check sorted.len == 3
    check sorted.find(1) < sorted.find(2)
    check sorted.find(1) < sorted.find(3)
    check sorted.find(2) < sorted.find(3)

echo "All tests completed!"
